File: app/src/main/java/org/lineageos/jelly/webview/WebViewExt.java
/*
 * Copyright (C) 2017 The LineageOS Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.lineageos.jelly.webview;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Paint;
import androidx.collection.ArrayMap;
import android.util.AttributeSet;
import android.util.Log;
import android.view.View;
import android.webkit.WebSettings;
import android.webkit.WebView;
import android.widget.ProgressBar;

import org.lineageos.jelly.ui.UrlBarController;
import org.lineageos.jelly.utils.PrefsUtils;
import org.lineageos.jelly.utils.UrlUtils;

import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class WebViewExt extends WebView {

    private static final String TAG = "WebViewExt";

    private static final String DESKTOP_DEVICE = "X11; Linux x86_64";
    private static final String DESKTOP_USER_AGENT_FALLBACK =
            "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/37.0.2049.0 Safari/537.36";

    private WebViewExtActivity mActivity;

    private String mMobileUserAgent;
    private String mDesktopUserAgent;

    private boolean mIncognito;
    private boolean mDesktopMode;
    private String mLastLoadedUrl;

    private final Map<String, String> mRequestHeaders = new ArrayMap<>();
    private static final String HEADER_DNT = "DNT";

    public WebViewExt(Context context) {
        super(context);
    }

    public WebViewExt(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    public WebViewExt(Context context, AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
    }

    @Override
    public void loadUrl(String url) {
        mLastLoadedUrl = url;
        followUrl(url);
    }

    void followUrl(String url) {
        String fixedUrl = UrlUtils.smartUrlFilter(url);
        if (fixedUrl != null) {
            super.loadUrl(fixedUrl, mRequestHeaders);
            return;
        }

        String templateUri = PrefsUtils.getSearchEngine(mActivity);
        fixedUrl = UrlUtils.getFormattedUri(templateUri, url);
        if (fixedUrl != null) {
            super.loadUrl(fixedUrl, mRequestHeaders);
        }
    }

    public String getLastLoadedUrl() {
        return mLastLoadedUrl;
    }

    private void setup() {
        getSettings().setJavaScriptEnabled(PrefsUtils.getJavascript(mActivity));
        getSettings().setJavaScriptCanOpenWindowsAutomatically(PrefsUtils.getJavascript(mActivity));
        getSettings().setGeolocationEnabled(PrefsUtils.getLocation(mActivity));
        getSettings().setSupportMultipleWindows(true);
        getSettings().setBuiltInZoomControls(true);
        getSettings().setDisplayZoomControls(false);
        getSettings().setAppCacheEnabled(!mIncognito);
        getSettings().setDatabaseEnabled(!mIncognito);
        getSettings().setDomStorageEnabled(!mIncognito);
        getSettings().setAppCachePath(mActivity.getDir("appcache", Context.MODE_PRIVATE).getPath());

        setOnLongClickListener(new OnLongClickListener() {
            boolean shouldAllowDownload;

            @Override
            public boolean onLongClick(View v) {
                HitTestResult result = getHitTestResult();
                switch (result.getType()) {
                    case HitTestResult.IMAGE_TYPE:
                    case HitTestResult.SRC_IMAGE_ANCHOR_TYPE:
                        shouldAllowDownload = true;
                    case HitTestResult.SRC_ANCHOR_TYPE:
                        mActivity.showSheetMenu(result.getExtra(), shouldAllowDownload);
                        shouldAllowDownload = false;
                        return true;
                }
                return false;
            }
        });

        setDownloadListener((url, userAgent, contentDisposition, mimeType, contentLength) ->
                mActivity.downloadFileAsk(url, contentDisposition, mimeType));

        // Mobile: Remove "wv" from the WebView's user agent. Some websites don't work
        // properly if the browser reports itself as a simple WebView.
        // Desktop: Generate the desktop user agent starting from the mobile one so that
        // we always report the current engine version.
        Pattern pattern = Pattern.compile("([^)]+ \\()([^)]+)(\\) .*)");
        Matcher matcher = pattern.matcher(getSettings().getUserAgentString());
        if (matcher.matches()) {
            String mobileDevice = matcher.group(2).replace("; wv", "");
            mMobileUserAgent = matcher.group(1) + mobileDevice + matcher.group(3);
            mDesktopUserAgent = matcher.group(1) + DESKTOP_DEVICE + matcher.group(3)
                    .replace(" Mobile ", " ");
            getSettings().setUserAgentString(mMobileUserAgent);
        } else {
            Log.e(TAG, "Couldn't parse the user agent");
            mMobileUserAgent = getSettings().getUserAgentString();
            mDesktopUserAgent = DESKTOP_USER_AGENT_FALLBACK;
        }

        if (PrefsUtils.getDoNotTrack(mActivity)) {
            mRequestHeaders.put(HEADER_DNT, "1");
        }
    }

    public void init(WebViewExtActivity activity, UrlBarController urlBarController,
                     ProgressBar progressBar, boolean incognito) {
        mActivity = activity;
        mIncognito = incognito;
        ChromeClient chromeClient = new ChromeClient(activity, incognito,
                urlBarController, progressBar);
        setWebChromeClient(chromeClient);
        setWebViewClient(new WebClient(urlBarController));
        setup();
    }

    public Bitmap getSnap() {
        measure(MeasureSpec.makeMeasureSpec(MeasureSpec.UNSPECIFIED, MeasureSpec.UNSPECIFIED),
                MeasureSpec.makeMeasureSpec(0, MeasureSpec.UNSPECIFIED));
        layout(0, 0, getMeasuredWidth(), getMeasuredHeight());
        setDrawingCacheEnabled(true);
        buildDrawingCache();
        int size = getMeasuredWidth() > getMeasuredHeight() ?
                getMeasuredHeight() : getMeasuredWidth();
        Bitmap bitmap = Bitmap.createBitmap(size, size, Bitmap.Config.ARGB_8888);
        Canvas canvas = new Canvas(bitmap);
        Paint paint = new Paint();
        int height = bitmap.getHeight();
        canvas.drawBitmap(bitmap, 0, height, paint);
        draw(canvas);
        return bitmap;
    }

    public boolean isIncognito() {
        return mIncognito;
    }

    public void setDesktopMode(boolean desktopMode) {
        mDesktopMode = desktopMode;
        WebSettings settings = getSettings();
        settings.setUserAgentString(desktopMode ? mDesktopUserAgent : mMobileUserAgent);
        settings.setUseWideViewPort(desktopMode);
        settings.setLoadWithOverviewMode(desktopMode);
        reload();
    }

    public boolean isDesktopMode() {
        return mDesktopMode;
    }

    Map<String, String> getRequestHeaders() {
        return mRequestHeaders;
    }
}

File: app/src/main/java/org/lineageos/jelly/webview/WebViewExt.kt
/*
 * SPDX-FileCopyrightText: 2020-2025 The LineageOS Project
 * SPDX-License-Identifier: Apache-2.0
 */

package org.lineageos.jelly.webview

import android.content.Context
import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Paint
import android.util.AttributeSet
import android.util.Log
import android.view.View
import android.webkit.WebView
import org.lineageos.jelly.js.JsManifest
import org.lineageos.jelly.js.JsSyncUrl
import org.lineageos.jelly.ui.UrlBarLayout
import org.lineageos.jelly.utils.SharedPreferencesExt
import org.lineageos.jelly.utils.UrlUtils
import java.util.regex.Pattern

class WebViewExt @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyle: Int = 0
) : WebView(context, attrs, defStyle) {
    private lateinit var activity: WebViewExtActivity
    val requestHeaders = mutableMapOf<String?, String?>()
    private var mobileUserAgent: String? = null
    private var desktopUserAgent: String? = null
    var isIncognito = false
        private set
    private var desktopMode = false
    var lastLoadedUrl: String? = null
        private set

    private val sharedPreferencesExt by lazy { SharedPreferencesExt(context) }

    override fun loadUrl(url: String) {
        lastLoadedUrl = url
        followUrl(url)
    }

    fun followUrl(url: String) {
        UrlUtils.smartUrlFilter(url)?.let {
            super.loadUrl(it, this.requestHeaders)
            return
        }
        val templateUri = sharedPreferencesExt.searchEngine
        super.loadUrl(UrlUtils.getFormattedUri(templateUri, url), this.requestHeaders)
    }

    private fun setup(urlBarLayout: UrlBarLayout) {
        settings.javaScriptEnabled = sharedPreferencesExt.javascriptEnabled
        settings.javaScriptCanOpenWindowsAutomatically = sharedPreferencesExt.javascriptEnabled
        settings.setGeolocationEnabled(sharedPreferencesExt.locationEnabled)
        settings.setSupportMultipleWindows(true)
        settings.builtInZoomControls = true
        settings.displayZoomControls = false
        settings.domStorageEnabled = !isIncognito
        setOnLongClickListener(object : OnLongClickListener {
            override fun onLongClick(v: View): Boolean {
                val result = hitTestResult
                result.extra?.let {
                    when (result.type) {
                        HitTestResult.IMAGE_TYPE, HitTestResult.SRC_IMAGE_ANCHOR_TYPE -> {
                            activity.showSheetMenu(it, true)
                            return true
                        }

                        HitTestResult.SRC_ANCHOR_TYPE -> {
                            activity.showSheetMenu(it, false)
                            return true
                        }

                        else -> {
                            return false
                        }
                    }
                }
                return false
            }
        })
        setDownloadListener { url: String?, userAgent: String?, contentDisposition: String?,
                              mimeType: String?, contentLength: Long ->
            activity.downloadFileAsk(url, userAgent, contentDisposition, mimeType, contentLength)
        }

        // Mobile: Remove "wv" and "Version/4.0" from the WebView's user agent.
        // Some websites don't work properly if the browser reports itself as a simple WebView.
        // Desktop: Generate the desktop user agent starting from the mobile one so that
        // we always report the current engine version.
        val pattern = Pattern.compile("([^)]+ \\()([^)]+)(\\) .*)")
        val matcher = pattern.matcher(settings.userAgentString)
        if (matcher.matches()) {
            val mobileDevice = matcher.group(2)!!.replace("; wv", "")
            mobileUserAgent = matcher.group(1)!! + mobileDevice + matcher.group(3)!!
                .replace(" Version/4.0 ", " ")
            desktopUserAgent = matcher.group(1)!! + DESKTOP_DEVICE + matcher.group(3)!!
                .replace(" Mobile ", " ")
                .replace(" Version/4.0 ", " ")
            settings.userAgentString = mobileUserAgent
        } else {
            Log.e(TAG, "Couldn't parse the user agent")
            mobileUserAgent = settings.userAgentString
            desktopUserAgent = DESKTOP_USER_AGENT_FALLBACK
        }
        if (sharedPreferencesExt.doNotTrackEnabled) {
            this.requestHeaders[HEADER_DNT] = "1"
        }

        if (settings.javaScriptEnabled) {
            addJavascriptInterface(
                JsSyncUrl(urlBarLayout, activity),
                JsSyncUrl.INTERFACE
            )
            addJavascriptInterface(
                JsManifest(activity),
                JsManifest.INTERFACE
            )
        }
    }

    fun init(
        activity: WebViewExtActivity, urlBarLayout: UrlBarLayout, incognito: Boolean
    ) {
        this.activity = activity
        isIncognito = incognito
        val chromeClient = ChromeClient(
            activity, incognito, urlBarLayout, sharedPreferencesExt
        )
        webChromeClient = chromeClient
        webViewClient = WebClient(urlBarLayout)
        setFindListener { activeMatchOrdinal, numberOfMatches, _ ->
            urlBarLayout.searchPositionInfo = Pair(activeMatchOrdinal, numberOfMatches)
        }
        urlBarLayout.onLoadUrlCallback = { loadUrl(it) }
        urlBarLayout.onStartSearchCallback = { findAllAsync(it) }
        urlBarLayout.onClearSearchCallback = { clearMatches() }
        urlBarLayout.onSearchPositionChangeCallback = { findNext(it) }
        setup(urlBarLayout)
    }

    val snap: Bitmap
        get() {
            measure(
                MeasureSpec.makeMeasureSpec(MeasureSpec.UNSPECIFIED, MeasureSpec.UNSPECIFIED),
                MeasureSpec.makeMeasureSpec(0, MeasureSpec.UNSPECIFIED)
            )
            layout(0, 0, measuredWidth, measuredHeight)
            val size = if (measuredWidth > measuredHeight) measuredHeight else measuredWidth
            val bitmap = Bitmap.createBitmap(size, size, Bitmap.Config.ARGB_8888)
            val canvas = Canvas(bitmap)
            val paint = Paint()
            val height = bitmap.height
            canvas.drawBitmap(bitmap, 0f, height.toFloat(), paint)
            draw(canvas)
            return bitmap
        }

    var isDesktopMode: Boolean
        get() = desktopMode
        set(desktopMode) {
            this.desktopMode = desktopMode
            val settings = settings
            settings.userAgentString = if (desktopMode) desktopUserAgent else mobileUserAgent
            settings.useWideViewPort = desktopMode
            settings.loadWithOverviewMode = desktopMode
            reload()
        }

    companion object {
        private const val TAG = "WebViewExt"
        private const val DESKTOP_DEVICE = "X11; Linux x86_64"
        private const val DESKTOP_USER_AGENT_FALLBACK =
            "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/131.0.0.0 Safari/537.36"
        private const val HEADER_DNT = "DNT"
    }
}
